File: day2/p1.hs
import System.IO (readFile)
import Data.Function

filename :: String
filename = "input.txt"

getTuples :: String -> [(String, Int)]
getTuples str =  fmap (\[w, i] -> (w, read i)) $ fmap words $ lines str

addTuples :: (Int, Int) -> (Int, Int) -> (Int, Int)
addTuples a b = (fst a + fst b, snd a + snd b)

getDim :: [(String, Int)] -> (Int, Int)
getDim [] = (0, 0)
getDim ((dir, mag):xs)
  | dir == "forward" = addTuples (mag, 0) (getDim xs)
  | dir == "up" = addTuples (0, -mag) (getDim xs)
  | dir == "down" = addTuples (0, mag) (getDim xs)

solve :: String -> (Int, Int)
solve str
  = getTuples str
  & getDim

main :: IO ()
main = do
  string <- readFile filename
  let solution = solve string
  print solution
  print $ (fst solution) * (snd solution)

File: day2/p1.py
#INPUT = "test.txt"
INPUT = "input.txt"

def get_final_pos(filename):
    h, d = 0, 0
    with open(filename) as fp:
        for line in fp:
            sp = line.split(" ")
            if sp[0] == 'forward':
                h += int(sp[1])
            elif sp[0] == 'down':
                d += int(sp[1])
            elif sp[0] == 'up':
                d -= int(sp[1])
    return h, d

if __name__ == "__main__":
    h, d = get_final_pos(INPUT)
    print(h, d, h*d)
